File: problems/二分查找/34.在排序数组中查找元素的第一个和最后一个位置.cpp
#include <iostream>
#include <vector>
#include <functional>
using namespace std;

vector<int> searchRange(vector<int>& nums, int target) {
    auto find_bound = [&](bool find_first) {
        int left = 0, right = nums.size() - 1, bound = -1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                bound = mid;
                if (find_first) right = mid - 1;
                else left = mid + 1;
            } else if (nums[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return bound;
    };
    return {find_bound(true), find_bound(false)};
}

int main() {
    // 测试用例
    vector<int> n1 = {5, 7, 7, 8, 8, 10};
    auto r1 = searchRange(n1, 8);
    cout << r1[0] << " " << r1[1] << endl;  // 预期: 3 4

    vector<int> n2 = {5, 7, 7, 8, 8, 10};
    auto r2 = searchRange(n2, 6);
    cout << r2[0] << " " << r2[1] << endl;  // 预期: -1 -1

    vector<int> n3 = {};
    auto r3 = searchRange(n3, 0);
    cout << r3[0] << " " << r3[1] << endl;  // 预期: -1 -1

    return 0;
}

File: problems/二分查找/34.在排序数组中查找元素的第一个和最后一个位置.py
#!/usr/bin/env python3
"""
34. 在排序数组中查找元素的第一个和最后一个位置
算法: 二分查找（分别找左边界和右边界）
时间复杂度: O(log n)
空间复杂度: O(1)
"""
def search_range(nums: list[int], target: int) -> list[int]:
    def find_bound(nums, target, find_first):
        left, right = 0, len(nums) - 1
        bound = -1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                bound = mid
                if find_first:
                    right = mid - 1
                else:
                    left = mid + 1
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return bound
    return [find_bound(nums, target, True), find_bound(nums, target, False)]


if __name__ == "__main__":
    print(search_range([5, 7, 7, 8, 8, 10], 8))       # 预期: [3, 4]
    print(search_range([5, 7, 7, 8, 8, 10], 6))       # 预期: [-1, -1]
    print(search_range([], 0))                         # 预期: [-1, -1]
